net-dns/yaddns/patches/yaddns-1.1.patchset

```diff
@@ -0,0 +1,237 @@
+         }
+ 
+         if(config_parse_file(cfg) != 0)
+diff --git a/src/ifaddr.cc b/src/ifaddr.cc
+new file mode 100644
+index 0000000..650d9a9
+--- /dev/null
++++ b/src/ifaddr.cc
+@@ -0,0 +1,28 @@
++#include "ifaddr.h"
++
++#include <NetworkInterface.h>
++#include <NetworkRoster.h>
++
++extern "C" {
++
++int getifaddr(const char *ifname, struct in_addr *addr)
++{
++	if(!ifname || ifname[0]=='\0')
++        {
++		return -1;
++        }
++
++	BNetworkInterface interface(ifname);
++	if (!interface.Exists()) return -1;
++
++	BNetworkInterfaceAddress address;
++	if (interface.GetAddressAt(0, address) != B_OK) return -1;
++
++	struct sockaddr sa = address.Address().SockAddr();
++	if (sa.sa_family != AF_INET) return -1;
++
++	*addr = ((struct sockaddr_in *)&sa)->sin_addr;
++	return 0;
++}
++
++}
+diff --git a/src/ifaddr.h b/src/ifaddr.h
+new file mode 100644
+index 0000000..ae723c3
+--- /dev/null
++++ b/src/ifaddr.h
+@@ -0,0 +1,17 @@
++#ifndef _IFADDR_H
++#define _IFADDR_H
++
++#include <netinet/in.h>
++#include <sys/socket.h>
++
++#ifdef __cplusplus
++extern "C" {
++#endif
++
++int getifaddr(const char *ifname, struct in_addr *addr);
++
++#ifdef __cplusplus
++}
++#endif
++
++#endif
+diff --git a/src/util.c b/src/util.c
+index 624ab3e..80fdb4b 100644
+--- a/src/util.c
++++ b/src/util.c
+@@ -8,6 +8,7 @@
+ #include <time.h>
+ #include <errno.h>
+ 
++#include "ifaddr.h"
+ #include "util.h"
+ #include "log.h"
+ 
+@@ -90,42 +91,12 @@ time_t util_getuptime(void)
+ 
+ int util_getifaddr(const char *ifname, struct in_addr *addr)
+ {
+-        /* SIOCGIFADDR struct ifreq *  */
+-	int s;
+-	struct ifreq ifr;
+-	int ifrlen;
+-
+-	if(!ifname || ifname[0]=='\0')
+-        {
+-		return -1;
+-        }
+-
+-	s = socket(PF_INET, SOCK_DGRAM, 0);
+-	if(s < 0)
+-	{
+-		log_error("socket(PF_INET, SOCK_DGRAM): %s",
+-                          strerror(errno));
+-		return -1;
+-	}
+-
+-        memset(&ifr, 0, sizeof(ifr));
+-        ifr.ifr_addr.sa_family = AF_INET; /* IPv4 IP address */
+-	strncpy(ifr.ifr_name, ifname, IFNAMSIZ-1);
+-	ifrlen = sizeof(ifr);
+-
+-	if(ioctl(s, SIOCGIFADDR, &ifr, &ifrlen) < 0)
+-	{
+-		log_error("ioctl(s, SIOCGIFADDR, ...): %s",
+-                          strerror(errno));
+-		close(s);
++	if (getifaddr(ifname, addr) == 0) {
++		return 0;
++	} else {
++		log_error("getifaddr(ifname, addr) failed");
+ 		return -1;
+ 	}
+-
+-	*addr = ((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr;
+-
+-	close(s);
+-
+-	return 0;
+ }
+ 
+ char *strdup_trim(const char *s)
+diff --git a/src/yaddns.c b/src/yaddns.c
+index 722d324..58a29d3 100644
+--- a/src/yaddns.c
++++ b/src/yaddns.c
+@@ -229,8 +229,6 @@ int main(int argc, char **argv)
+                 goto exit_clean;
+         }
+ 
+-        sig_blockall();
+-
+         sigemptyset(&unblocked);
+ 
+ 	/* config */
+@@ -250,12 +248,24 @@ int main(int argc, char **argv)
+         /* daemonize ? */
+         if(cfg.daemonize)
+         {
+-                if(daemon(0, 0) < 0)
```

Comment:
We have [`daemon()`][0], define `_BSD_SOURCE` to use.

[0]: http://xref.plausible.coop/source/xref/haiku/src/libs/bsd/daemon.c#24

Reply:
I've just checked that this is working fine:
```diff
 BUILD()
 {
 	./autogen.sh
+	CFLAGS="-DBSD_SOURCE" CXXFLAGS="-D_BSD_SOURCE" LDFLAGS="-lbsd" \
 	LIBS="-Wl,--as-needed -lbnetapi -lnetwork" runConfigure \
 		--omit-dirs binDir ./configure --bindir $commandBinDir
 
 	make $jobArgs
 }
```
With this change, the part of the patchset which patches src/yaddns.c can be completely removed.

BTW, it also builds and passes all the tests without the `CXXFLAGS="-D_BSD_SOURCE"`.
(I haven't checked, however, if the resulting runtimes are identical.)

Reply:
You should append `-lbsd` to `LIBS` (additional libraries to link) instead. For `-D_BSD_SOURCE`, I'd recommend to put it into `CPPFLAGS` (preprocessor flags), which is used by both C and C++ compilers.